brave/src/main/java/brave/internal/propagation/InjectorFactory.java

```diff
@@ -0,0 +1,351 @@
+        + "}";
+  }
+
+  static final class DeferredInjector<R> implements TraceContext.Injector<R> {
+    final Setter<R, String> setter;
+    final InjectorFactory injectorFactory;
+
+    DeferredInjector(Setter<R, String> setter, InjectorFactory injectorFactory) {
+      this.setter = setter;
+      this.injectorFactory = injectorFactory;
+    }
+
+    @Override public void inject(TraceContext context, R request) {
+      if (request instanceof Request) {
+        switch (((Request) request).spanKind()) {
+          case CLIENT:
+            injectorFactory.clientInjectorFunction.inject(setter, context, request);
+            return;
+          case PRODUCER:
+            injectorFactory.producerInjectorFunction.inject(setter, context, request);
+            return;
+          case CONSUMER:
+            injectorFactory.consumerInjectorFunction.inject(setter, context, request);
+          default: // SERVER is nonsense as it cannot be injected
+        }
+      }
+      injectorFactory.injectorFunction.inject(setter, context, request);
+    }
+
+    @Override public int hashCode() {
+      int h = 1000003;
+      h ^= setter.hashCode();
+      h *= 1000003;
+      h ^= injectorFactory.hashCode();
+      return h;
+    }
+
+    @Override public boolean equals(Object o) {
+      if (o == this) return true;
+      if (!(o instanceof DeferredInjector)) return false;
+      DeferredInjector<?> that = (DeferredInjector<?>) o;
+      return setter.equals(that.setter) && injectorFactory.equals(that.injectorFactory);
+    }
+
+    @Override public String toString() {
+      return "DeferredInjector{setter=" + setter + ", injectorFactory=" + injectorFactory + "}";
+    }
+  }
+
+  static final class Injector<R> implements TraceContext.Injector<R> {
+    final InjectorFunction injectorFunction;
+    final Setter<R, String> setter;
+
+    Injector(Setter<R, String> setter, InjectorFunction injectorFunction) {
+      this.injectorFunction = injectorFunction;
+      this.setter = setter;
+    }
+
+    @Override public void inject(TraceContext context, R request) {
+      injectorFunction.inject(setter, context, request);
+    }
+
+    @Override public int hashCode() {
+      int h = 1000003;
+      h ^= setter.hashCode();
+      h *= 1000003;
+      h ^= injectorFunction.hashCode();
+      return h;
+    }
+
+    @Override public boolean equals(Object o) {
+      if (o == this) return true;
+      if (!(o instanceof Injector)) return false;
+      Injector<?> that = (Injector<?>) o;
+      return setter.equals(that.setter) && injectorFunction.equals(that.injectorFunction);
+    }
+
+    @Override public String toString() {
+      return "Injector{setter=" + setter + ", injectorFunction=" + injectorFunction + "}";
+    }
+  }
+
+  static InjectorFunction injectorFunction(InjectorFunction... injectorFunctions) {
+    if (injectorFunctions == null) throw new NullPointerException("injectorFunctions == null");
```

Comment:
Don't think `Arrays.asList` does so probably want to check all the elements aren't null too

Reply:
sure thing